### ISSUE TITLE
Add Optimistic Mirror Descent (OMD)

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#46947]
## 🚀 Feature

Add Optimistic Mirror Descent (OMD) to PyTorch's [optimizers](https://pytorch.org/docs/stable/optim.html).

OMD is described in Algorithm 1 of paper 1 and Algorithm 3 of paper 2.

Papers:
1. [Training GANs with Optimism](https://arxiv.org/abs/1711.00141) (2017)
2. [Optimistic mirror descent in saddle-point problems: Going the extra (gradient) mile](https://arxiv.org/abs/1807.02629) (2018)

There seems to be a PyTorch implementation [here](https://discuss.pytorch.org/t/optimistic-mirror-descent-in-saddle-point-problems-adam-optimizer-modification/26928) and a Tensorflow implementation [here](https://gist.github.com/EndingCredits/b5f35e84df10d46cfa716178d9c862a3).

## Motivation

As described in the above papers, OMD works better than vanilla MD for solving saddle-point problems.

These saddle point problems arise in adversarial settings, e.g. training [GANs](https://en.wikipedia.org/wiki/Generative_adversarial_network) and multi-agent reinforcement learning.
